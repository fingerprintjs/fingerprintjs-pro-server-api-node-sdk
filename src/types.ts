import { components, paths } from './generatedApiTypes'

export enum Region {
  EU = 'EU',
  AP = 'AP',
  Global = 'Global',
}

export enum AuthenticationMode {
  AuthHeader = 'AuthHeader',
  QueryParameter = 'QueryParameter',
}

/**
 * Options for FingerprintJS server API client
 */
export interface Options {
  /**
   * Secret API key
   */
  apiKey: string
  /**
   * Region of the FingerprintJS service server
   */
  region?: Region
  /**
   * Authentication mode
   * Optional, default value is AuthHeader
   */
  authenticationMode?: AuthenticationMode

  /**
   * Optional fetch implementation
   * */
  fetch?: typeof fetch
}

/**
 * More info: https://dev.fingerprintjs.com/docs/server-api#query-parameters
 */
export type VisitorHistoryFilter = paths['/visitors/{visitor_id}']['get']['parameters']['query']

export type TooManyRequestsError = {
  status: 429
  /**
   * How many seconds to wait before retrying
   */
  retryAfter: number
}

/**
 * More info: https://dev.fingerprintjs.com/docs/server-api#response
 */
export type VisitorsResponse = paths['/visitors/{visitor_id}']['get']['responses']['200']['content']['application/json']
export type VisitorsResponse403 =
  paths['/visitors/{visitor_id}']['get']['responses']['403']['content']['application/json'] & {
    status: 403
  }
export type VisitorsResponse429 =
  paths['/visitors/{visitor_id}']['get']['responses']['429']['content']['application/json']

export type DeleteVisit404Response =
  paths['/visitors/{visitor_id}']['delete']['responses']['404']['content']['application/json']

export type DeleteVisit403Response =
  paths['/visitors/{visitor_id}']['delete']['responses']['403']['content']['application/json']

export type DeleteVisit400Response =
  paths['/visitors/{visitor_id}']['delete']['responses']['400']['content']['application/json'] & {
    status: 400
  }

export type CommonResponse429 = components['schemas']['ErrorCommon429Response']

export type VisitorsError = WithResponse<VisitorsResponse403 | VisitorsResponse429>

export type EventResponse = paths['/events/{request_id}']['get']['responses']['200']['content']['application/json']
export type EventResponse403 = paths['/events/{request_id}']['get']['responses']['403']['content']['application/json']
export type EventResponse404 = paths['/events/{request_id}']['get']['responses']['404']['content']['application/json']

type WithResponse<T> = T & {
  response: Response
}

/**
 * More info: https://dev.fingerprintjs.com/docs/webhooks#identification-webhook-object-format
 */
export type VisitWebhook = components['schemas']['WebhookVisit']
