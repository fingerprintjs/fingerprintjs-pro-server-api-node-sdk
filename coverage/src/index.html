
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for src</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../prettify.css" />
    <link rel="stylesheet" href="../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../index.html">All files</a> src</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">94.85% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>166/175</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">83.33% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>45/54</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">96.55% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>28/29</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">94.79% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>164/173</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line high'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="index.ts"><a href="index.ts.html">index.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="9" class="abs high">9/9</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="9" class="abs high">9/9</td>
	</tr>

<tr>
	<td class="file high" data-value="responseUtils.ts"><a href="responseUtils.ts.html">responseUtils.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="6" class="abs high">6/6</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="6" class="abs high">6/6</td>
	</tr>

<tr>
	<td class="file high" data-value="sealedResults.ts"><a href="sealedResults.ts.html">sealedResults.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="39" class="abs high">39/39</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="9" class="abs high">9/9</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="6" class="abs high">6/6</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="39" class="abs high">39/39</td>
	</tr>

<tr>
	<td class="file high" data-value="serverApiClient.ts"><a href="serverApiClient.ts.html">serverApiClient.ts</a></td>
	<td data-value="89.55" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 89%"></div><div class="cover-empty" style="width: 11%"></div></div>
	</td>
	<td data-value="89.55" class="pct high">89.55%</td>
	<td data-value="67" class="abs high">60/67</td>
	<td data-value="63.63" class="pct medium">63.63%</td>
	<td data-value="22" class="abs medium">14/22</td>
	<td data-value="90" class="pct high">90%</td>
	<td data-value="10" class="abs high">9/10</td>
	<td data-value="89.55" class="pct high">89.55%</td>
	<td data-value="67" class="abs high">60/67</td>
	</tr>

<tr>
	<td class="file high" data-value="types.ts"><a href="types.ts.html">types.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="7" class="abs high">7/7</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="7" class="abs high">7/7</td>
	</tr>

<tr>
	<td class="file high" data-value="urlUtils.ts"><a href="urlUtils.ts.html">urlUtils.ts</a></td>
	<td data-value="96.96" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 96%"></div><div class="cover-empty" style="width: 4%"></div></div>
	</td>
	<td data-value="96.96" class="pct high">96.96%</td>
	<td data-value="33" class="abs high">32/33</td>
	<td data-value="91.66" class="pct high">91.66%</td>
	<td data-value="12" class="abs high">11/12</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="7" class="abs high">7/7</td>
	<td data-value="96.77" class="pct high">96.77%</td>
	<td data-value="31" class="abs high">30/31</td>
	</tr>

<tr>
	<td class="file medium" data-value="utils.ts"><a href="utils.ts.html">utils.ts</a></td>
	<td data-value="75" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 75%"></div><div class="cover-empty" style="width: 25%"></div></div>
	</td>
	<td data-value="75" class="pct medium">75%</td>
	<td data-value="4" class="abs medium">3/4</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="75" class="pct medium">75%</td>
	<td data-value="4" class="abs medium">3/4</td>
	</tr>

<tr>
	<td class="file high" data-value="webhook.ts"><a href="webhook.ts.html">webhook.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="10" class="abs high">10/10</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="10" class="abs high">10/10</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at 2025-06-04T13:30:03.627Z
            </div>
        <script src="../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../sorter.js"></script>
        <script src="../block-navigation.js"></script>
    </body>
</html>
    