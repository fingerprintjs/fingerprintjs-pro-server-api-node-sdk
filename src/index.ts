export * from './urlUtils'
export * from './serverApiClient'
export * from './types'
export * from './sealedResults'
export * from './errors/unsealError'
export * from './webhook'
export * from './errors/apiErrors'
