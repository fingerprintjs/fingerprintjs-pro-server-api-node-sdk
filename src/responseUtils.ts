import { SdkError } from './errors/apiErrors'
import { toError } from './utils'

export async function copyResponseJson(response: Response) {
  try {
    return await response.clone().json()
  } catch (e) {
    throw new SdkError('Failed to parse JSON response', response, toError(e))
  }
}
