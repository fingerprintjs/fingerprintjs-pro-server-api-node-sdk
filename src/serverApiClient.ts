import { getDeleteVisitorDataUrl, getEventUrl, getVisitorsUrl } from './urlUtils'
import { AuthenticationMode, EventResponse, Options, Region, VisitorHistoryFilter, VisitorsResponse } from './types'
import { copyResponseJson } from './responseUtils'
import {
  ApiError,
  CommonError429,
  DeleteVisit400Error,
  DeleteVisit403Error,
  DeleteVisit404Error,
  EventError403,
  EventError404,
  VisitorsError403,
  VisitorsError429,
} from './errors/apiErrors'

export class FingerprintJsServerApiClient {
  public readonly region: Region

  public readonly apiKey: string

  public readonly authenticationMode: AuthenticationMode

  protected readonly fetch: typeof fetch

  protected static readonly DEFAULT_RETRY_AFTER = 1

  /**
   * FingerprintJS server API client used to fetch data from FingerprintJS
   * @constructor
   * @param {Options} options - Options for FingerprintJS server API client
   */
  constructor(options: Readonly<Options>) {
    if (!options.apiKey) {
      throw Error('Api key is not set')
    }

    this.region = options.region ?? Region.Global
    this.apiKey = options.apiKey
    this.authenticationMode = options.authenticationMode ?? AuthenticationMode.AuthHeader // Default auth mode is AuthHeader
    this.fetch = options.fetch ?? fetch
  }

  /**
   * Retrieves a specific identification event with the information from each activated product â€” Identification and all active [Smart signals](https://dev.fingerprint.com/docs/smart-signals-overview).
   *
   * @param requestId - identifier of the event
   *
   * @returns {Promise<EventResponse>} - promise with event response. For more information, see the [Server API documentation](https://dev.fingerprint.com/reference/getevent).
   *
   * @example
   * ```javascript
   * client
   *  .getEvent('<requestId>')
   *  .then((result) => console.log(result))
   *  .catch((err) => {
   *    if (isEventError(err)) {
   *      // You can also access the raw response
   *      console.log(err.response)
   *      console.log(`error ${err.statusCode}: `, err.message)
   *    } else {
   *      console.log('unknown error: ', err)
   *    }
   *  })
   * ```
   * */
  public async getEvent(requestId: string): Promise<EventResponse> {
    if (!requestId) {
      throw new TypeError('requestId is not set')
    }

    const url =
      this.authenticationMode === AuthenticationMode.QueryParameter
        ? getEventUrl(requestId, this.region, this.apiKey)
        : getEventUrl(requestId, this.region)

    const headers = this.getHeaders()

    const response = await this.fetch(url, {
      method: 'GET',
      headers,
    })

    const jsonResponse = await copyResponseJson(response)

    if (response.status === 200) {
      return jsonResponse as EventResponse
    }

    switch (response.status) {
      case 403:
        throw new EventError403(jsonResponse, response)

      case 404:
        throw new EventError404(jsonResponse, response)

      default:
        throw ApiError.unknown(response)
    }
  }

  /**
   * Delete data by visitor ID
   * Request deleting all data associated with the specified visitor ID. This API is useful for compliance with privacy regulations. All delete requests are queued:
   * Recent data (10 days or newer) belonging to the specified visitor will be deleted within 24 hours. * Data from older (11 days or more) identification events  will be deleted after 90 days.
   * If you are interested in using this API, please [contact our support team](https://fingerprint.com/support/) to activate it for you. Otherwise, you will receive a 403.
   *
   * @param visitorId The [visitor ID](https://dev.fingerprint.com/docs/js-agent#visitorid) you want to delete.*
   *
   * @return {Promise<void>} Promise that resolves when the deletion request is successfully queued
   *
   * @example
   * ```javascript
   * client
   *   .deleteVisitorData('<visitorId>')
   *   .then(() => {
   *     // Data deletion request was successfully queued
   *   })
   *   .catch((error) => {
   *     if (isDeleteVisitorError(error)) {
   *       console.log(error.statusCode, error.message)
   *     }
   *   })
   * ```
   */
  public async deleteVisitorData(visitorId: string): Promise<void> {
    if (!visitorId) {
      throw TypeError('VisitorId is not set')
    }

    const url =
      this.authenticationMode === AuthenticationMode.QueryParameter
        ? getDeleteVisitorDataUrl(this.region, visitorId, this.apiKey)
        : getDeleteVisitorDataUrl(this.region, visitorId)

    const headers = this.getHeaders()

    const response = await this.fetch(url, {
      method: 'DELETE',
      headers,
    })

    if (response.status === 200) {
      return
    }

    const jsonResponse = await copyResponseJson(response)

    switch (response.status) {
      case 429:
        throw new CommonError429(jsonResponse, response)

      case 404:
        throw new DeleteVisit404Error(jsonResponse, response)

      case 403:
        throw new DeleteVisit403Error(jsonResponse, response)

      case 400:
        throw new DeleteVisit400Error(jsonResponse, response)

      default:
        throw ApiError.unknown(response)
    }
  }

  /**
   * Retrieves event history for the specific visitor using the given filter, returns a promise with visitor history response.
   *
   * @param {string} visitorId - Identifier of the visitor
   * @param {VisitorHistoryFilter} filter - Visitor history filter
   * @param {string} filter.limit - limit scanned results
   * @param {string} filter.request_id - filter visits by `requestId`.
   * @param {string} filter.linked_id - filter visits by your custom identifier.
   * @param {string} filter.paginationKey - use `paginationKey` to get the next page of results.   When more results are available (e.g., you requested 200 results using `limit` parameter, but a total of 600 results are available), the `paginationKey` top-level attribute is added to the response. The key corresponds to the `requestId` of the last returned event. In the following request, use that value in the `paginationKey` parameter to get the next page of results:
   *
   *   1. First request, returning most recent 200 events: `GET api-base-url/visitors/:visitorId?limit=200`
   *   2. Use `response.paginationKey` to get the next page of results: `GET api-base-url/visitors/:visitorId?limit=200&paginationKey=1683900801733.Ogvu1j`
   *
   *   Pagination happens during scanning and before filtering, so you can get less visits than the `limit` you specified with more available on the next page. When there are no more results available for scanning, the `paginationKey` attribute is not returned.
   * @example
   * ```javascript
   * client
   *   .getVisitorHistory('<visitorId>', { limit: 1 })
   *   .then((visitorHistory) => {
   *     console.log(visitorHistory)
   *   })
   *   .catch((error) => {
   *     if (isVisitorsError(error)) {
   *       console.log(error.statusCode, error.message)
   *       if (error.status === 429) {
   *         retryLater(error.retryAfter) // Needs to be implemented on your side
   *       }
   *     }
   *   })
   * ```
   */
  public async getVisitorHistory(visitorId: string, filter?: VisitorHistoryFilter): Promise<VisitorsResponse> {
    if (!visitorId) {
      throw TypeError('VisitorId is not set')
    }

    const url =
      this.authenticationMode === AuthenticationMode.QueryParameter
        ? getVisitorsUrl(this.region, visitorId, filter, this.apiKey)
        : getVisitorsUrl(this.region, visitorId, filter)
    const headers = this.getHeaders()

    const response = await this.fetch(url, {
      method: 'GET',
      headers,
    })

    const jsonResponse = await copyResponseJson(response)

    if (response.status === 200) {
      return jsonResponse as VisitorsResponse
    }

    switch (response.status) {
      case 403:
        throw new VisitorsError403(jsonResponse, response)

      case 429:
        throw new VisitorsError429(jsonResponse, response)

      default:
        throw ApiError.unknown(response)
    }
  }

  private getHeaders() {
    return this.authenticationMode === AuthenticationMode.AuthHeader ? { 'Auth-API-Key': this.apiKey } : undefined
  }
}
