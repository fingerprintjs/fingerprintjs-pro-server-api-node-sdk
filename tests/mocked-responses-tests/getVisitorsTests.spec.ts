import { VisitorHistoryFilter, Region } from '../../src/types'
import { FingerprintJsServerApiClient } from '../../src/serverApiClient'
import visitorsWithoutFilterResponse from './mocked-responses-data/visitors-without-filter-response.json'
import visitorsResponseWithRequestId from './mocked-responses-data/visitors-with-request-id.json'
import visitorsResponseWithRequestIdLinkedId from './mocked-responses-data/visitors-with-request-id-linked-id.json'
import visitorsResponseWithLinkedIdLimit from './mocked-responses-data/visitors-with-linked-id-limit.json'
import visitorsWithLimitBefore from './mocked-responses-data/visitors-with-limit-before.json'

jest.spyOn(global, 'fetch')

describe('[Mocked response] Get Visitors', () => {
  const apiKey = 'dummy_api_key'
  const existingVisitorId = 'TaDnMBz9XCpZNuSzFUqP'
  const existingRequestId = '1626550679751.cVc5Pm'
  const existingLinkedId = 'makma'

  const client = new FingerprintJsServerApiClient({ region: Region.EU, apiKey: apiKey })

  test('without filter', async () => {
    ;(fetch as unknown as jest.Mock).mockReturnValue(
      Promise.resolve(new Response(JSON.stringify(visitorsWithoutFilterResponse)))
    )

    const response = await client.getVisitorHistory(existingVisitorId)
    expect(response).toMatchSnapshot()
  })

  test('with request_id filter', async () => {
    ;(fetch as unknown as jest.Mock).mockReturnValue(
      Promise.resolve(new Response(JSON.stringify(visitorsResponseWithRequestId)))
    )

    const filter: VisitorHistoryFilter = { request_id: existingRequestId }
    const response = await client.getVisitorHistory(existingVisitorId, filter)
    expect(response).toMatchSnapshot()
  })

  test('with request_id and linked_id filter', async () => {
    ;(fetch as unknown as jest.Mock).mockReturnValue(
      Promise.resolve(new Response(JSON.stringify(visitorsResponseWithRequestIdLinkedId)))
    )

    const filter: VisitorHistoryFilter = {
      request_id: existingRequestId,
      linked_id: existingLinkedId,
    }
    const response = await client.getVisitorHistory(existingVisitorId, filter)
    expect(response).toMatchSnapshot()
  })

  test('with linked_id and limit filter', async () => {
    ;(fetch as unknown as jest.Mock).mockReturnValue(
      Promise.resolve(new Response(JSON.stringify(visitorsResponseWithLinkedIdLimit)))
    )

    const filter: VisitorHistoryFilter = { linked_id: existingLinkedId, limit: 5 }
    const response = await client.getVisitorHistory(existingVisitorId, filter)
    expect(response).toMatchSnapshot()
  })

  test('with limit and before', async () => {
    ;(fetch as unknown as jest.Mock).mockReturnValue(
      Promise.resolve(new Response(JSON.stringify(visitorsWithLimitBefore)))
    )

    const filter: VisitorHistoryFilter = { limit: 4, before: 1626538505244 }
    const response = await client.getVisitorHistory(existingVisitorId, filter)
    expect(response).toMatchSnapshot()
  })

  test('403 error', async () => {
    ;(fetch as unknown as jest.Mock).mockReturnValue(
      Promise.resolve(
        new Response(
          JSON.stringify({
            error: 'Forbidden',
          }),
          {
            status: 403,
          }
        )
      )
    )
    await expect(client.getVisitorHistory(existingVisitorId)).rejects.toMatchObject({
      status: 403,
      error: 'Forbidden',
    })
  })

  test('429 error', async () => {
    ;(fetch as unknown as jest.Mock).mockReturnValue(
      Promise.resolve(
        new Response(
          JSON.stringify({
            error: 'Too Many Requests',
          }),
          {
            status: 429,
            headers: { 'Retry-after': '10' },
          }
        )
      )
    )
    await expect(client.getVisitorHistory(existingVisitorId)).rejects.toMatchObject({
      status: 429,
      error: 'Too Many Requests',
      retryAfter: 10,
    })
  })

  test('429 error with empty retry-after header', async () => {
    ;(fetch as unknown as jest.Mock).mockReturnValue(
      Promise.resolve(
        new Response(
          JSON.stringify({
            error: 'Too Many Requests',
          }),
          {
            status: 429,
          }
        )
      )
    )
    await expect(client.getVisitorHistory(existingVisitorId)).rejects.toMatchObject({
      status: 429,
      error: 'Too Many Requests',
      retryAfter: 1,
    })
  })

  test('Error with bad JSON', async () => {
    ;(fetch as unknown as jest.Mock).mockReturnValue(
      Promise.resolve(
        new Response('(Some bad JSON)', {
          status: 404,
        })
      )
    )
    await expect(client.getVisitorHistory(existingVisitorId)).rejects.toMatchObject({
      status: 0,
    })
  })
})
