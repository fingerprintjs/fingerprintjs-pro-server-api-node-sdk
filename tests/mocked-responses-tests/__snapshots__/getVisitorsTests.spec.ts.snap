// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Mocked response] Get Visitors with limit and before 1`] = `
{
  "lastTimestamp": 1655373953086,
  "paginationKey": "1655373953086.DDlfmP",
  "visitorId": "AcxioeQKffpXF8iGQK3P",
  "visits": [
    {
      "browserDetails": {
        "botProbability": 0,
        "browserFullVersion": "102.0.5005",
        "browserMajorVersion": "102",
        "browserName": "Chrome",
        "device": "Other",
        "os": "Mac OS X",
        "osVersion": "10.15.7",
        "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/102.0.5005.61 Safari/537.36",
      },
      "confidence": {
        "score": 1,
      },
      "firstSeenAt": {
        "global": "2022-02-04T11:31:20Z",
        "subscription": "2022-02-04T11:31:20Z",
      },
      "incognito": false,
      "ip": "82.118.30.68",
      "ipLocation": {
        "accuracyRadius": 1000,
        "city": {
          "name": "Prague",
        },
        "continent": {
          "code": "EU",
          "name": "Europe",
        },
        "country": {
          "code": "CZ",
          "name": "Czechia",
        },
        "latitude": 50.0805,
        "longitude": 14.467,
        "postalCode": "130 00",
        "subdivisions": [
          {
            "isoCode": "10",
            "name": "Hlavni mesto Praha",
          },
        ],
        "timezone": "Europe/Prague",
      },
      "lastSeenAt": {
        "global": "2022-06-16T10:03:00.912Z",
        "subscription": "2022-06-16T10:03:00.912Z",
      },
      "requestId": "1655373953086.DDlfmP",
      "tag": {},
      "time": "2022-06-16T10:05:53Z",
      "timestamp": 1655373953094,
      "url": "https://dashboard.fingerprint.com/",
      "visitorFound": true,
    },
  ],
}
`;

exports[`[Mocked response] Get Visitors with linked_id and limit filter 1`] = `
{
  "lastTimestamp": 1655373953086,
  "paginationKey": "1655373953086.DDlfmP",
  "visitorId": "AcxioeQKffpXF8iGQK3P",
  "visits": [
    {
      "browserDetails": {
        "botProbability": 0,
        "browserFullVersion": "102.0.5005",
        "browserMajorVersion": "102",
        "browserName": "Chrome",
        "device": "Other",
        "os": "Mac OS X",
        "osVersion": "10.15.7",
        "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/102.0.5005.61 Safari/537.36",
      },
      "confidence": {
        "score": 1,
      },
      "firstSeenAt": {
        "global": "2022-02-04T11:31:20Z",
        "subscription": "2022-02-04T11:31:20Z",
      },
      "incognito": false,
      "ip": "82.118.30.68",
      "ipLocation": {
        "accuracyRadius": 1000,
        "city": {
          "name": "Prague",
        },
        "continent": {
          "code": "EU",
          "name": "Europe",
        },
        "country": {
          "code": "CZ",
          "name": "Czechia",
        },
        "latitude": 50.0805,
        "longitude": 14.467,
        "postalCode": "130 00",
        "subdivisions": [
          {
            "isoCode": "10",
            "name": "Hlavni mesto Praha",
          },
        ],
        "timezone": "Europe/Prague",
      },
      "lastSeenAt": {
        "global": "2022-06-16T10:03:00.912Z",
        "subscription": "2022-06-16T10:03:00.912Z",
      },
      "requestId": "1655373953086.DDlfmP",
      "tag": {},
      "time": "2022-06-16T10:05:53Z",
      "timestamp": 1655373953094,
      "url": "https://dashboard.fingerprint.com/",
      "visitorFound": true,
    },
  ],
}
`;

exports[`[Mocked response] Get Visitors with request_id and linked_id filter 1`] = `
{
  "lastTimestamp": 1655373953086,
  "paginationKey": "1655373953086.DDlfmP",
  "visitorId": "AcxioeQKffpXF8iGQK3P",
  "visits": [
    {
      "browserDetails": {
        "botProbability": 0,
        "browserFullVersion": "102.0.5005",
        "browserMajorVersion": "102",
        "browserName": "Chrome",
        "device": "Other",
        "os": "Mac OS X",
        "osVersion": "10.15.7",
        "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/102.0.5005.61 Safari/537.36",
      },
      "confidence": {
        "score": 1,
      },
      "firstSeenAt": {
        "global": "2022-02-04T11:31:20Z",
        "subscription": "2022-02-04T11:31:20Z",
      },
      "incognito": false,
      "ip": "82.118.30.68",
      "ipLocation": {
        "accuracyRadius": 1000,
        "city": {
          "name": "Prague",
        },
        "continent": {
          "code": "EU",
          "name": "Europe",
        },
        "country": {
          "code": "CZ",
          "name": "Czechia",
        },
        "latitude": 50.0805,
        "longitude": 14.467,
        "postalCode": "130 00",
        "subdivisions": [
          {
            "isoCode": "10",
            "name": "Hlavni mesto Praha",
          },
        ],
        "timezone": "Europe/Prague",
      },
      "lastSeenAt": {
        "global": "2022-06-16T10:03:00.912Z",
        "subscription": "2022-06-16T10:03:00.912Z",
      },
      "requestId": "1655373953086.DDlfmP",
      "tag": {},
      "time": "2022-06-16T10:05:53Z",
      "timestamp": 1655373953094,
      "url": "https://dashboard.fingerprint.com/",
      "visitorFound": true,
    },
  ],
}
`;

exports[`[Mocked response] Get Visitors with request_id filter 1`] = `
{
  "lastTimestamp": 1655373953086,
  "paginationKey": "1655373953086.DDlfmP",
  "visitorId": "AcxioeQKffpXF8iGQK3P",
  "visits": [
    {
      "browserDetails": {
        "botProbability": 0,
        "browserFullVersion": "102.0.5005",
        "browserMajorVersion": "102",
        "browserName": "Chrome",
        "device": "Other",
        "os": "Mac OS X",
        "osVersion": "10.15.7",
        "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/102.0.5005.61 Safari/537.36",
      },
      "confidence": {
        "score": 1,
      },
      "firstSeenAt": {
        "global": "2022-02-04T11:31:20Z",
        "subscription": "2022-02-04T11:31:20Z",
      },
      "incognito": false,
      "ip": "82.118.30.68",
      "ipLocation": {
        "accuracyRadius": 1000,
        "city": {
          "name": "Prague",
        },
        "continent": {
          "code": "EU",
          "name": "Europe",
        },
        "country": {
          "code": "CZ",
          "name": "Czechia",
        },
        "latitude": 50.0805,
        "longitude": 14.467,
        "postalCode": "130 00",
        "subdivisions": [
          {
            "isoCode": "10",
            "name": "Hlavni mesto Praha",
          },
        ],
        "timezone": "Europe/Prague",
      },
      "lastSeenAt": {
        "global": "2022-06-16T10:03:00.912Z",
        "subscription": "2022-06-16T10:03:00.912Z",
      },
      "requestId": "1655373953086.DDlfmP",
      "tag": {},
      "time": "2022-06-16T10:05:53Z",
      "timestamp": 1655373953094,
      "url": "https://dashboard.fingerprint.com/",
      "visitorFound": true,
    },
  ],
}
`;

exports[`[Mocked response] Get Visitors without filter 1`] = `
{
  "lastTimestamp": 1655373953086,
  "paginationKey": "1655373953086.DDlfmP",
  "visitorId": "AcxioeQKffpXF8iGQK3P",
  "visits": [
    {
      "browserDetails": {
        "botProbability": 0,
        "browserFullVersion": "102.0.5005",
        "browserMajorVersion": "102",
        "browserName": "Chrome",
        "device": "Other",
        "os": "Mac OS X",
        "osVersion": "10.15.7",
        "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/102.0.5005.61 Safari/537.36",
      },
      "confidence": {
        "score": 1,
      },
      "firstSeenAt": {
        "global": "2022-02-04T11:31:20Z",
        "subscription": "2022-02-04T11:31:20Z",
      },
      "incognito": false,
      "ip": "82.118.30.68",
      "ipLocation": {
        "accuracyRadius": 1000,
        "city": {
          "name": "Prague",
        },
        "continent": {
          "code": "EU",
          "name": "Europe",
        },
        "country": {
          "code": "CZ",
          "name": "Czechia",
        },
        "latitude": 50.0805,
        "longitude": 14.467,
        "postalCode": "130 00",
        "subdivisions": [
          {
            "isoCode": "10",
            "name": "Hlavni mesto Praha",
          },
        ],
        "timezone": "Europe/Prague",
      },
      "lastSeenAt": {
        "global": "2022-06-16T10:03:00.912Z",
        "subscription": "2022-06-16T10:03:00.912Z",
      },
      "requestId": "1655373953086.DDlfmP",
      "tag": {},
      "time": "2022-06-16T10:05:53Z",
      "timestamp": 1655373953094,
      "url": "https://dashboard.fingerprint.com/",
      "visitorFound": true,
    },
  ],
}
`;
